third_party/binutils/test_decenc/drdecode_decenc_x86.c

```diff
@@ -0,0 +1,3786 @@
+    test_s((byte *)test_sse2avx_s_asm);
+#        endif
+#        ifdef test_sse3_s
+    dr_printf("test_sse3_s:\n");
+    test_s((byte *)test_sse3_s_asm);
+#        endif
+#        ifdef test_sse4_1_s
+    dr_printf("test_sse4_1_s:\n");
+    test_s((byte *)test_sse4_1_s_asm);
+#        endif
+#        ifdef test_sse4_2_s
+    dr_printf("test_sse4_2_s:\n");
+    test_s((byte *)test_sse4_2_s_asm);
+#        endif
+#        ifdef test_ssemmx2_s
+    dr_printf("test_ssemmx2_s:\n");
+    test_s((byte *)test_ssemmx2_s_asm);
+#        endif
+#        ifdef test_ssse3_s
+    dr_printf("test_ssse3_s:\n");
+    test_s((byte *)test_ssse3_s_asm);
+#        endif
+#        ifdef test_string_ok_s
+    dr_printf("test_string_ok_s:\n");
+    test_s((byte *)test_string_ok_s_asm);
+#        endif
+#        ifdef test_sub_s
+    dr_printf("test_sub_s:\n");
+    test_s((byte *)test_sub_s_asm);
+#        endif
+#        ifdef test_suffix_s
+    dr_printf("test_suffix_s:\n");
+    test_s((byte *)test_suffix_s_asm);
+#        endif
+#        ifdef test_svme_s
+    dr_printf("test_svme_s:\n");
+    test_s((byte *)test_svme_s_asm);
+#        endif
+#        ifdef test_tbm_s
+    dr_printf("test_tbm_s:\n");
+    test_s((byte *)test_tbm_s_asm);
+#        endif
+#        ifdef test_tlsd_s
+    dr_printf("test_tlsd_s:\n");
+    test_s((byte *)test_tlsd_s_asm);
+#        endif
+#        ifdef test_tlsnopic_s
+    dr_printf("test_tlsnopic_s:\n");
+    test_s((byte *)test_tlsnopic_s_asm);
+#        endif
+#        ifdef test_tlspic_s
+    dr_printf("test_tlspic_s:\n");
+    test_s((byte *)test_tlspic_s_asm);
+#        endif
+#        ifdef test_vaes_s
+    dr_printf("test_vaes_s:\n");
+    test_s((byte *)test_vaes_s_asm);
+#        endif
+#        ifdef test_vex_lig_2_s
+    dr_printf("test_vex_lig_2_s:\n");
+    test_s((byte *)test_vex_lig_2_s_asm);
+#        endif
+#        ifdef test_vpclmulqdq_s
+    dr_printf("test_vpclmulqdq_s:\n");
+    test_s((byte *)test_vpclmulqdq_s_asm);
+#        endif
+#        ifdef test_waitpkg_s
+    dr_printf("test_waitpkg_s:\n");
+    test_s((byte *)test_waitpkg_s_asm);
+#        endif
+#        ifdef test_wbnoinvd_s
+    dr_printf("test_wbnoinvd_s:\n");
+    test_s((byte *)test_wbnoinvd_s_asm);
+#        endif
+#        ifdef test_white_s
+    dr_printf("test_white_s:\n");
+    test_s((byte *)test_white_s_asm);
+#        endif
+#        ifdef test_xop_s
+    dr_printf("test_xop_s:\n");
+    test_s((byte *)test_xop_s_asm);
+#        endif
+#        ifdef test_xmmhi64_s
+    dr_printf("test_xmmhi64_s:\n");
+    test_s((byte *)test_xmmhi64_s_asm);
+#        endif
+#        ifdef test_xop32reg_s
+    dr_printf("test_xop32reg_s:\n");
+    test_s((byte *)test_xop32reg_s_asm);
+#        endif
+#        ifdef test_xsave_s
+    dr_printf("test_xsave_s:\n");
+    test_s((byte *)test_xsave_s_asm);
+#        endif
+#        ifdef test_xsavec_s
+    dr_printf("test_xsavec_s:\n");
+    test_s((byte *)test_xsavec_s_asm);
+#        endif
+# 
```

Comment:
This approach seems to require a lot of work to edit.
If I want to add or remove a test, I have to go and edit 7 different places in shared files:

1. The list in suite/tests/CMakeLists.txt.
1. The function declaration at the top of this file.
1. The ifdef around the print and test_s call.
1. The print of the test name.
1. The call to test_s().
1. Another ifdef down here.
1. The include of the .asm file.

That is a lot of places storing the unique name, for each of 400 separate names.

I then have to create 3 new files: the .asm, the .expect, and the .runcmp.  The .runcmp can easily be eliminated, but that is a lot of places where these names are stored.

I only looked at a couple .asm files, but many seem pretty short.  In fact, they are not that much longer than the 3 lines here for including them.  My proposal is to inline all the .asm files here and have a single .expect file.  That eliminates all of the duplication of these separate test names and IMHO makes it simpler both to read the code and to maintain it.  We can have comments labeling the test boundaries in the expect file: so we'd have a subtest name once here in the code and once in the expect file.

I don't see a big benefit from having these as completely separate tests: only in seeing what failed, but that is a rare occurrence.  The downsides outweigh that IMHO: it is awkward dealing with 400 names all duplicated in various places, and 1200 (!) separate files, and building and running 400 separate binaries takes a lot longer than building just one.

I prefer fewer files when we're talking about hundreds or thousands of them, but an alternative could be separate code and expect files, and just get rid of all the declarations and ifdefs and concatenate everything together here and in the single master expect file.

Reply:
I wanted to keep the original separation from binutils, such that if they add something, one can easily see what they've added. Yes, you have to edit a few places, but this is just a one time thing that (will-) almost never happen(s). Re-factoring this like this will take me another few hours potentially. The advantage is also to be able to easily disable the parts that are currently not working, just by editing them in suite/tests/CMakeLists.txt (#1312, #3577, #3578, #3581).

Reply:
> I wanted to keep the original separation from binutils, such that if they add something, one can easily see what they've added.

I don't see how the organization in our repo relates to that: simply look at the binutils version control.  The names would still be in comments in any case which is just as easy to check as a list of files.

If these were straight from binutils then I could see the argument: but these are heavily edited.  Or if they were truly separate: but they are already bundled and #included into a single .c file anyway.  Once we have that, why build and run as separate tests?  That part seems to not offer advantages.  Agreed?

The separate files is then a further issue that perhaps should go to a wider audience then if we disagree.

> Yes, you have to edit a few places, but this is just a one time thing that (will-) almost never happen(s). Re-factoring this like this will take me another few hours potentially. The advantage is also to be able to easily disable the parts that are currently not working, just by editing them in suite/tests/CMakeLists.txt (#1312, #3577, #3578, #3581).

Disabling is going to be rarer than reading, adding, editing.

Reply:
See below.